### ISSUE TITLE
[feature request] improving the workflow for editing racket/gui 

### ISSUE THREAD

**Issue by Bogdanp** [greghendershott/racket-mode#602]
This is an extremely niche request so feel free to ignore it.

I've been working on adding more native widgets to `racket/gui`. The problem is it's hard to work on `racket/gui` (or `racket/gui`-adjacent stuff (such as [this](https://github.com/Bogdanp/racket-gui-extra))) in `racket-mode`. It would be nice if `racket-run-module-at-point` could restart the `racket-mode` process when it detects changes to certain user-specified paths. That way, I could manually add those paths (eg. `~/path/to/gui-lib/mred, ~/path/to/gui-extra-lib`) to my config, and it would simplify my workflow.

**Comment by greghendershott** [comment]
I'm not sure if it's niche, but even if it were, I'd like to support you continuing to do such good things.

As a first approximation, if you were to `M-x racket-start-back-end` whenever one of these directories changes, I think that's what you want to happen (except you needing to do so manually).

Is that right?

If so, then probably I'd look first at how to do filesystem change notifications in Emacs.

(If there's no good portable way to handle that, in Emacs, then I might look at having the Racket back end use Racket's [filesystem change notification](https://docs.racket-lang.org/reference/Filesystem.html#%28part._filesystem-change%29). And the back end notifies the front end. Or maybe the back end just exits, since the front end should detect and restart it automatically when next needed.  Not sure yet. Just thinking out loud.)

**Comment by capfredf** [comment]
I'd like to chime in: this is really a nice-to-have feature. I do what @greghendershott suggests or (c-x c-v on the open file) roughly 20 times on a daily basis.

**Comment by greghendershott** [comment]
Emacs has [`file-notify-add-watch`](https://www.gnu.org/software/emacs/manual/html_node/elisp/File-Notifications.html) which I haven't used before but seems fine.

A couple tiny wrinkles (not a big deal):

- IIUC for directories `file-notify-add-watch` only notifies changes to files in that directory -- not recursively for subdirectories. Assuming you'll need that (?) I'll need to create multiple watches, one for each subdirectory recursively.

- Now that Racket Mode supports multiple back ends, it will need to know which one you want to restart. So, the config would probably be some sort of association list from directories to back end names (with `/` being the name of the default local back end).

**Comment by Bogdanp** [comment]
> As a first approximation, if you were to M-x racket-start-back-end whenever one of these directories changes, I think that's what you want to happen (except you needing to do so manually).

I hadn't realized I could use `racket-start-back-end`, so I was making my life harder by stopping the process via `list-processes` every time 😅. 

> Emacs has [file-notify-add-watch](https://www.gnu.org/software/emacs/manual/html_node/elisp/File-Notifications.html) which I haven't used before but seems fine.

I've tried using it to set up a handler that runs `racket-start-back-end`, and that works, though it's a bit noisy because I save frequently, so I then tried to have the handler set a flag that the backend ought to be restarted and advised `racket-run-module-at-point` to call `racket-start-back-end` before running when the flag is set, and that works sometimes, but sometimes there seems to be some sort of a race. For example:

```elisp

  (defvar bp-racket-mode-pending-reload nil)
  (defun bp-racket-mode-reload-backend (event)
    (unless (string-match "/\\.#" (caddr event))
      (setq bp-racket-mode-pending-reload t)))

  (file-notify-add-watch "/Users/bogdan/sandbox/racket-gui-extra/gui-extra-lib/gui/extra/private/cocoa" '(change) #'bp-racket-mode-reload-backend)
  (defadvice racket-run-module-at-point (before bp-racket-reload activate)
    (when bp-racket-mode-pending-reload
      (setq bp-racket-mode-pending-reload nil)
      (racket-start-back-end)))

```

Fails with 

```
racket-back-end-/ command exception:
run: current-repl-msg-chan was #f; current-session-id=#f
```

about half the time.

>  IIUC for directories file-notify-add-watch only notifies changes to files in that directory -- not recursively for subdirectories. Assuming you'll need that (?) I'll need to create multiple watches, one for each subdirectory recursively.

Yes, ideally I would point it at the root directory for one of these libs and have it recursively discover the directories.

**Comment by greghendershott** [comment]
Yes, after working on this for awhile (before seeing your comment just now), I think there are enough wrinkles for this to be tricky for you to do as a self-service tweak.

For example I also saw what you did with the command error. This seems due to the old REPL process needing some time to finish terminating. (If you do the commands manually, `racket-start-back-end`, then `racket-run`, there's enough time between them for it to work reliably.)

I just pushed a commit [89e7f8b](https://github.com/greghendershott/racket-mode/commit/89e7f8b28f77af282a2f7836b4f06d6a63edb808) to a topic branch `issue-602` which "works for me".

If you don't already have a `racket-add-back-end` in your init file, then you need one now for this, but just something like:

```elisp
(racket-add-back-end
  "/" ;the default back end for all local files
  :restart-watch-directories '("/Users/bogdan/sandbox/racket-gui-extra/gui-extra-lib/gui/extra/private/cocoa"))
```

I'm going to sleep on this a bit. However, if using a Racket Mode branch here is convenient for you, and you'd like to try it, please let me know.

**Comment by Bogdanp** [comment]
I just tried the branch and it works well. I probably won't have much time to test this out until the weekend, but I'll let you know if I run into any issues once I do. Thanks!

**Comment by greghendershott** [comment]
Cool; sounds good so far.

One thing I wasn't sure about is the UX of the `y-or-n-prompt`. On the one hand maybe that will be annoying if it happens a lot. On the other hand I like the feedback and chance to confirm. So I'm curious what you will think, after you have a chance to live with it awhile.

**Comment by Bogdanp** [comment]
I think this is great so far. I don't mind the prompt at all. After using the functionality for a bit today, I've gotten used to and anticipate the prompt so I end up hitting `C-c C-k` followed by `y` without really thinking about it. 

One additional problem I'm running into while working on `gui-extra` is that, like `gui-lib`, it must only be instantiated once per process. This comes up when I try to work on examples: the first time I load an example module, everything works great, but usually by the time I try to re-load that same module things crash.  The problem goes away if I add `gui-extra` to `racket/gui.rkt`:

```diff
diff --git a/racket/gui.rkt b/racket/gui.rkt
index 07fb707..fdd59ae 100644
--- a/racket/gui.rkt
+++ b/racket/gui.rkt
@@ -29,7 +29,8 @@
         (system* (find-executable-path (find-system-path 'exec-file))
                  "-e" "(require racket/gui/base)"))
   (with-handlers ([exn:fail? void])
-    (dynamic-require 'racket/gui/base #f)))
+    (dynamic-require 'racket/gui/base #f)
+    (dynamic-require 'racket/gui/extra #f)))
 
 ;; #301: On Windows, show then hide an initial frame.
 (when (and (gui-available?)
@@ -52,7 +53,8 @@
   ;; whether a given user program `require`s it. They might require it
   ;; in the REPL. See also issue #555.
   (when (gui-available?)
-    (namespace-attach-module our-ns 'racket/gui/base new-ns))
+    (namespace-attach-module our-ns 'racket/gui/base new-ns)
+    (namespace-attach-module our-ns 'racket/gui/extra new-ns))
 
   ;; Avoid potential problem (IIUC because Racket structs are
   ;; generative) with file/convertible by attaching the same instance
```

Would it be possible/would it make sense for there to be a configurable list of singleton modules per backend?

**Comment by greghendershott** [comment]
> I think this is great so far. I don't mind the prompt at all. After using the functionality for a bit today, I've gotten used to and anticipate the prompt so I end up hitting C-c C-k followed by y without really thinking about it.

Glad to hear that!

> One additional problem I'm running into while working on gui-extra is that, like gui-lib, it must only be instantiated once per process. This comes up when I try to work on examples: the first time I load an example module, everything works great, but usually by the time I try to re-load that same module things crash.

Hmm.

1. Do we know why? Is that because `gui-extra` itself imposes this, e.g. has some once-per-process resource use via FFI or whatever?

2. Does the same happen if you try this in Dr Racket?

---

I'm 100% OK with adding configuration for more "singleton modules". But I'd hate to overlook some other/better way to handle whatever is going on.

**Comment by greghendershott** [comment]
> Do we know why? Is that because gui-extra itself imposes this, e.g. has some once-per-process resource use via FFI or whatever?

Taking a super quick  (and probably insufficiently caffeinated) peek at `racket/gui/extra` source: It looks like the spirit here is that effectively you're using and enhancing the `mred` internals.

Maybe it's not necessary for me to understand this too deeply, or, think about some more-general solution for once-per-process modules?  You seem to be doing some very cool experimental hacking tied to the one part of Racket (AFAIK) with this constraint.

So probably it's fine, in the same spirit, just to add some super-specific feature for "additional modules to handle the same as `racket/gui/base`"?

In other words, just do exactly what you requested without worrying about it too much? :smile:

**Comment by Bogdanp** [comment]
Yup! I think it's also equally fine for racket-mode not to support anything like this at this time, since it's easy enough for me to run a modified racket-mode. Maybe this functionality is only worth adding if `gui-extra` ever becomes useful or if more things with similar constraints pop up in time.

**Comment by greghendershott** [comment]
Thank you for the feedback!

In that case:

- I think I'll only merge the `:restart-watch-directories` change, and close this issue, for now.

- As for "extra singleton modules": Someday if someone else has a similar or "adjacent" need (or if you eventually find it a PITA to keep patching Racket Mode like this) that could be a tipping point where I revisit this.

**Comment by Bogdanp** [comment]
Sounds good! Thanks!

**Comment by greghendershott** [comment]
Merged!
